MaxText/layers/pipeline.py

```diff
@@ -0,0 +1,332 @@
+
+import jax
+import jax.ad_checkpoint
+import numpy as np
+from jax import numpy as jnp
+from jax import tree_map
+from flax.core import meta
+from flax import linen as nn
+import common_types
+import functools
+from typing import Any
+
+class Pipeline(nn.Module):
+  """Module that implements pipelining across stages.
+  
+  This module will loop over microbatches and execute the main body with a vmap for both the inputs and weights.
+  This will produce a pipeline pattern if the stage dimension is sharded.
+
+  Does not yet support circular pipelines. Multiple
+  layers per stage are used when a module that executes multiple layers per stage is passed as the layers input.
+
+  Attributes:
+    config: Importantly contains num_pipeline_microbatches.
+    layers: A module instance that each stage can execute. It can either be a single layer such as a LlamaDecoderLayer instance
+      or scanned/looped set of decoder layers to execute multiple layers per stage.
+    mesh:  The device mesh of the system.
+    remat_policy: Remat policy to use for the loop iterations
+  """
+  
+  config: common_types.Config
+  layers: nn.Module # The name of this property (layers) is reflected in the state pytree and thus also checkpoints.
+  mesh: common_types.Mesh
+  remat_policy: Any = None
+
+  def setup(self):
+    self.num_stages = self.config.ici_pipeline_parallelism * self.config.dcn_pipeline_parallelism
+    self.microbatch_size = self.config.global_batch_size_to_train_on // self.config.num_pipeline_microbatches
+    microbatches_per_stage = self.config.num_pipeline_microbatches // self.num_stages
+    self.microbatches_per_stage = microbatches_per_stage
+
+  def init_states(self, inputs):
+    '''Initialize components of state: state_io, shift
+        Assumes input has already been reshaped into microbatches: [num_micro_batches, micro_batch_size, sequence, embed]
+
+        Returns a dictionary with properties
+          shift: zeros shape [num_stages, micro_size, sequence, embed]
+          state_io: reshaped inputs [num_stages, microbatches/stages, micro_size, sequence, embed]
+          loop_iteration: scalar set initially to 0.  
+    '''
+
+    # Shift is used to rotate the output of each pipeline into the input of the next
+    # shift has shape [num_stages, micro_size, sequence, embed]
+    shift = jnp.zeros((self.num_stages,) + inputs.shape[1:], dtype=inputs.dtype)
+    shift = nn.with_logical_constraint(shift, ("activation_stage", "activation_batch", "activation_length", "activation_embed"),rules=self.config.logical_axis_rules,mesh=self.mesh)
+
+    # state_io (state input output) at first holds all of the input batches, but also will hold the outputs as the pipeline runs/finishes
+    # state_io has shape [num_stages, microbatches/stages, micro_size, sequence, embed]
+    state_io = jnp.reshape(inputs, (self.num_stages, self.microbatches_per_stage) + inputs.shape[1:])
```

Comment:
It might be possible to have only one microbatch dimension for state_io, but the separation allows for easy manipulation of the desired sharding pattern (shard by stages) with minimal resharding required - The pattern is that each iteration we need to push a microbatch out of state_io and also one into state_io from stages - here we are guaranteed that the microbatch at the top of state_io is on the correct device(s) (stage 0 device(s)), and the microbatch to be pushed to the end is also at the correct device (corresponding to the last stage), and we only need to collective permute microbatches_per_stage at each iteration instead of all of them. This may be possible with just a single axis but would require complex bookkeeping/indexing